File: 0739-daily-temperatures/0739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        ms = []
        ans = []
        for i in range(len(temperatures)-1, -1, -1):
            while ms and ms[-1][0] <= temperatures[i]:
                ms.pop()
            if ms:
                ans.append(ms[-1][1]-i)
            else:
                ans.append(0)
            ms.append((temperatures[i], i))
        return ans[::-1]
        

File: 0739-daily-temperatures/0739-daily-temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int len = temperatures.length;
        Stack<int[]> stack = new Stack<>();
        int arr[] = new int[len];
        
        for(int i = len-1 ; i>=0 ; i -- ) {
   
            while(!stack.isEmpty() && stack.peek()[0] <= temperatures[i]) {
                stack.pop();
            }
            
            if (!stack.isEmpty())
                arr[i] = stack.peek()[1] - i;
                
            stack.push(new int[] {temperatures[i], i});
            }
        
        return arr;
    }
}